### PR TITLE
Support for more types

### PR DESCRIPTION
**Pull request by carldebilly**
- More immutable types
- support for collections/dictionaries in equality generation
- minor fixes

### CODE REVIEW

File: src/Uno.CodeGen/EqualityGenerator.cs [hunk 1]
```diff
@@ -408,13 +412,50 @@ private void GenerateEqualLogic(INamedTypeSymbol typeSymbol, IndentedStringBuild
...
 					{
-						builder.AppendLineInvariant($"return false; // {member.Name} not equal");
+						// Extract mode from attribute, or default following the type of the collection (if ordered)
+						var collectionAttribute = member.FindAttribute(_equalityCollectionAttributeSymbol);
+						var mode = (int)(collectionAttribute?.ConstructorArguments[0].Value
+							?? (isOrdered ? 1 : 2));
+
+						var comparer = mode == 1
+							? (isReadonlyCollection ? "SortedReadonlyCollectionEqualityComparer" : "SortedCollectionEqualityComparer")
+							: (isReadonlyCollection ? "UnsortedReadonlyCollectionEqualityComparer" : "UnsortedCollectionEqualityComparer");
```
Comment (jeromelaban) [inline src/Uno.CodeGen/EqualityGenerator.cs]: Maybe use constants.

File: src/Uno.CodeGen/EqualityGenerator.cs [hunk 2]
```diff
@@ -539,8 +582,8 @@ private void GenerateHashLogic(INamedTypeSymbol typeSymbol, IndentedStringBuilde
 								builder.AppendLineInvariant(
 									$"#warning Type `{type.GetDisplayFriendlyName()}` of member `{member.Name}` " +
 									"doesn't implements .GetHashCode(): it won't be used for hash computation. " +
```
Comment (jeromelaban) [inline src/Uno.CodeGen/EqualityGenerator.cs]: "Does not implement"

File: src/Uno.CodeGen/EqualityGenerator.cs [hunk 3]
```diff
@@ -539,8 +582,8 @@ private void GenerateHashLogic(INamedTypeSymbol typeSymbol, IndentedStringBuilde
...
-									$"If you can change the type {type}, you should use a custom hash method or " +
-									"a custom comparer.");
+									$"If you can't change the type {type}, you should use a custom hash method or " +
+									"a custom comparer. Alternatively, use something else for hash computation.");
```
Comment (jeromelaban) [inline src/Uno.CodeGen/EqualityGenerator.cs]: "something else" give an example ?

File: src/Uno.Equality/CollectionComparerMode.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+// ******************************************************************
+namespace Uno
+{
+	/// <summary>
+	/// Use to qualify the <see cref="EqualityCollectionAttribute"/> attribute.
+	/// </summary>
+	public enum CollectionComparerMode
+	{
+		/// <summary>
+		/// Use a comparer which check the ordering in the collection.
```
Comment (jeromelaban) [inline src/Uno.Equality/CollectionComparerMode.cs]: Use a comparer which **checks for** the ordering in the collection.

File: src/Uno.Equality/CollectionComparerMode.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+		/// </summary>
+		Sorted = 1,
+
+		/// <summary>
+		/// Use a comparer which allows a different ordering between collections.
```
Comment (jeromelaban) [inline src/Uno.Equality/CollectionComparerMode.cs]: Use a comparer which allows **for** a different ordering between collections.

File: src/Uno.CodeGen/EqualityGenerator.cs (hunk 1 above)
Comment (carldebilly) [reply src/Uno.CodeGen/EqualityGenerator.cs]: Done :-)

File: src/Uno.CodeGen/Uno.CodeGen.csproj [hunk 6]
```diff
@@ -25,6 +25,8 @@
 
 	<ItemGroup Condition="'$(TargetFramework)'!='net46'">
 		<Compile Remove="**\*.cs" />
+		<None Include="**\*.cs" />
```
[the remark this answers is not in the document]
Comment (carldebilly) [reply src/Uno.CodeGen/Uno.CodeGen.csproj]: Yep it's better.

**Pull request merged**